serverless-functions/src/functions/features/callback-and-voicemail/studio/wait-experience.protected.js

```diff
@@ -0,0 +1,340 @@
+  // we are still in the waitUrl TwiML execution - and it's not possible to use the <Record> verb in here. We piggyback on the same approach for callbacks,
+  // though technically these could be handled entirely in the waitUrl TwiML
+  const redirectUrl = `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=${mode}&CallSid=${callSid}&enqueuedTaskSid=${taskSid}`;
+  const result = await VoiceOperations.updateCall({
+    context,
+    callSid,
+    params: { method: "POST", url: redirectUrl },
+    attempts: 0,
+  });
+  const { success, status } = result;
+
+  if (success) {
+    //  Cancel (update) the task with handy attributes for reporting
+    await cancelTask(context, task, cancelReason);
+  } else {
+    console.error(
+      `Failed to update call ${callSid} with new TwiML. Status: ${status}`
+    );
+    twiml.say(
+      sayOptions,
+      "Sorry, we were unable to perform this operation. Please remain on the line."
+    );
+    twiml.redirect(
+      `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=main-wait-loop&CallSid=${callSid}&enqueuedTaskSid=${taskSid}&skipGreeting=true`
+    );
+    return twiml;
+  }
+  return "";
+}
+
+exports.handler = async function (context, event, callback) {
+  const domain = `https://${context.DOMAIN_NAME}`;
+  const twiml = new Twilio.twiml.VoiceResponse();
+
+  // Retrieve options
+  const { sayOptions, holdMusicUrl } = options;
+
+  const {
+    Digits,
+    CallSid,
+    enqueuedWorkflowSid,
+    mode,
+    enqueuedTaskSid,
+    skipGreeting,
+  } = event;
+
+  let message = "";
+
+  switch (mode) {
+    case "initialize":
+      // Initial logic to find the associated task for the call, and propagate it through to the rest of the TwiML execution
+      // If the lookup fails to find the task, the remaining TwiML logic will not offer any callback or voicemail options.
+      const enqueuedTask = await getPendingTaskByCallSid(
+        context,
+        CallSid,
+        enqueuedWorkflowSid
+      );
+
+      const redirectUrlNoTaskFound = `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=main-wait-loop&CallSid=${CallSid}`;
+
+      if (enqueuedTask == null) {
+        // Log an error for our own debugging purposes, but don't fail the call
+        console.error(
+          `Failed to find the pending task with callSid: ${CallSid}. This is potentially due to higher call volume than the API query had accounted for.`
+        );
+        twiml.redirect(redirectUrlNoTaskFound);
+      } else {
+        twiml.redirect(
+          redirectUrlNoTaskFound +
+            (enqueuedTask ? `&enqueuedTaskSid=${enqueuedTask.sid}` : "")
+        );
+      }
+      return callback(null, twiml);
+
+    case "main-wait-loop":
+      if (skipGreeting !== "true") {
+        const initGreeting =
+          "Please wait while we direct your call to the next available specialist.";
```

Comment:
I think this (and other `say` strings) should be part of the `options` object, since customers will probably want to customize them, just like the other options.